lib/cosmos/packets/structure.rb

```diff
@@ -327,7 +332,37 @@ def buffer=(buffer)
     # Take the structure mutex to ensure the buffer does not change while you perform activities
     def synchronize
       @mutex ||= Mutex.new
-      @mutex.synchronize {|| yield }
+      @mutex.synchronize {|| yield}
+    end
+
+    # Take the structure mutex to ensure the buffer does not change while you perform activities
+    # This versions allows reads to happen if a top level function has already taken the mutex
```

Comment:
By top level I mean the method that starts the calling sequence.   This is just within Structure/Packet for now but they could be subclassed.   It's hard to screw up and not notice because you will try to take the mutex twice and throw and exception.